tools/prow-auto-bumper/getversion_test.go

```diff
@@ -0,0 +1,368 @@
+				},
+				"p": []versions{
+					{"a-b-y", "h-i-y", "y"},
+				},
+			},
+			versions{"a-b", "h-i", ""},
+		},
+	}
+
+	for _, data := range datas {
+		images := data.images
+		pv := PRVersions{
+			images: images,
+		}
+		if vs := pv.getDominantVersions(); vs != data.vs {
+			log.Fatalf("get dominant versions for '%v', want: '%v', got: '%v'", images, data.vs, vs)
+		}
+	}
+}
+
+func TestParseChangelist(t *testing.T) {
+	datas := []struct {
+		patches []string
+		images  map[string][]versions
+	}{
+		{
+			[]string{},
+			map[string][]versions{},
+		},
+		{
+			[]string{
+				"- image: gcr.io/k8s-foofoo/bar:va-b",
+				"+ image: gcr.io/k8s-foofoo/bar:vh-i",
+			},
+			map[string][]versions{
+				"gcr.io/k8s-foofoo/bar": []versions{
+					{"va-b", "vh-i", ""},
+				},
+			},
+		},
+		{
+			[]string{
+				"- image: gcr.io/k8s-foofoo/bar:va-b",
+				"+ image: gcr.io/k8s-foofoo/bar:vh-i",
+				"- image: gcr.io/k8s-foofoo/bar:va-b-x",
+				"+ image: gcr.io/k8s-foofoo/bar:vh-i-x",
+			},
+			map[string][]versions{
+				"gcr.io/k8s-foofoo/bar": []versions{
+					{"va-b", "vh-i", ""},
+					{"va-b-x", "vh-i-x", "x"},
+				},
+			},
+		},
+		{
+			[]string{
+				`- image: gcr.io/k8s-foofoo/bar:va-b
+				+ image: gcr.io/k8s-foofoo/bar:vh-i
+				- image: gcr.io/k8s-foofoo/bar:va-b-x
+				+ image: gcr.io/k8s-foofoo/bar:vh-i-x`,
+			},
+			map[string][]versions{
+				"gcr.io/k8s-foofoo/bar": []versions{
+					{"va-b", "vh-i", ""},
+					{"va-b-x", "vh-i-x", "x"},
+				},
+			},
+		},
+		{
+			[]string{
+				"- image: gcr.io/k8s-foofoo/bar:va-b",
+				"+ image: gcr.io/k8s-foofoo/bar:vh-i",
+				"- image: gcr.io/k8s-barbar/baz:vc-d",
+				"+ image: gcr.io/k8s-barbar/baz:vj-k",
+			},
+			map[string][]versions{
+				"gcr.io/k8s-foofoo/bar": []versions{
+					{"va-b", "vh-i", ""},
+				},
+				"gcr.io/k8s-barbar/baz": []versions{
+					{"vc-d", "vj-k", ""},
+				},
+			},
+		},
+	}
+
+	for _, data := range datas {
+		fgc := fakeghutil.NewFakeGithubClient()
+		fc := &Client{fgc}
+		PR, err := fgc.CreatePullRequest(fakeOrg, fakeRepo, "user:head", "base", "title", "body")
+		if nil != err {
+			t.Fatalf("Create PR in %s/%s, want: no error, got: '%v'", fakeOrg, fakeRepo, err)
+		}
+		pv := PRVersions{
+			images: make(map[string][]versions),
+			PR:     PR,
+		}
+		for i, patch := range data.patches {
+			SHA := strconv.Itoa(i)
+			filename := fmt.Sprintf("file_%d", i)
+			fgc.AddCommitToPullRequest(fakeOrg, fakeRepo, *pv.PR.Number, SHA)
+			fgc.AddFileToCommit(fakeOrg, fakeRepo, SHA, filename, patch)
+		}
+
+		pv.parseChangelist(fc)
+		if eq := reflect.DeepEqual(pv.images, data.images); !eq {
+			t.Fatalf("parsing PR with changes '%v', want: '%v', got: '%v'",
+				data.patches, data.images, pv.images)
+		}
+	}
+}
+
+func TestGetBestVersion(t *testing.T) {
+	type PRInfo struct {
+		delta      int // hours before current date
+		state      ghutil.PullRequestState
+		oldVersion string
+		newVersion string
+	}
+	datas := []struct {
+		PIs        []PRInfo
```

Comment:
I suggest using `PRInfos` instead of `PIs`.

Reply:
Also `dominantVersions`.

Reply:
done